Updates for Moriond

blinkseb opened the pull request:
See commits for details. I used tight id for both electrons and muons, but this need to be discussed, especially in the case of electrons where there's also the HLT safe ID and extra Impact parameter cuts we can choose to apply or not.

Commit:
Update HLT path

 - Ask for both DZ and non-DZ version for DoubleMuons
 - Only ask for Mu8 for ElMu

Commit:
Use latest JER factors

Commit:
Use POG IDs instead of HWW

swertz commented:
Ok, let's wait before merging then...

OlivierBondu commented:
so, chatting a bit IRL we discussed:
- medium for electrons (there are SF available, so why not)
- tight for muons (medium is run-dependent == a pain == avoid)

any objection ?

blinkseb commented:
Good for me, I'll update the PR asap

Commit:
Add DZ filter efficiencies for HLT paths

blinkseb commented:
Update using new WP for electrons. I also added DZ filter efficiencies directly in the analyzer, so no need to apply it at plotting stage.

HLT efficiencies must be updated before merging this.

blinkseb commented:
I just added a new commit to enable splitting of jec uncertainty by sources. Since this adds *a lot* of new stuff to the tree, I also cleaned a bit the code to:
 - Keep only *1* ll candidate passing our requirement
 - Keep only *1* llmetjj candidate (the first one), with a flag `btag_MM` to know if it passes b-tagging requirement or not

This allows to free some spaces since anyway we only use the first candidate, and we sometimes store up to 50! I also commented a bit of stuff around which is never used afterwards to again free some spaces. I'd anyway appreciate a review by some external eyes to ensure nothing is broken (Seb already run the plotter over this output and everything seems ok but who knows!)

OlivierBondu commented on interface/HHAnalyzer.h:
```diff
@@ -105,20 +109,21 @@ class HHAnalyzer: public Framework::Analyzer {
         // global event stuff (selected objects multiplicity)
-        BRANCH(nJets, unsigned int);
+        ONLY_NOMINAL_BRANCH(nJets, unsigned int);
         BRANCH(nJetsL, unsigned int);
```
should be `ONLY_NOMINAL_BRANCH` as well ?

blinkseb replied:
Actually no it's used by Seb for the DY BDT :)

swertz replied:
Indeed, but with your changes it's now equivalent to `nJets`, so I can switch to that, and then we do not need to store `nJetsL`, `nBJetsL`, `nBJetsT`, `nMuonsT`, `nElectronsM`, since you're keeping only the objects passing our IDs?

blinkseb replied:
If I remove these we have no longer a way to know the number of leptons passing the selection other than looping over the `hh_leptons` collection and counting.

swertz replied:
Sorry, indeed for the leptons it matters, but for the jets because of https://github.com/cp3-llbb/HHAnalysis/pull/125/files#diff-6cdb42586ea6220d1c53fa646556a54cR615 there is no difference between nJets and nJetsL?

blinkseb replied:
Indeed I was just lazy :)

OlivierBondu commented on plugins/HHAnalyzer.cc:
```diff
@@ -327,61 +329,34 @@ void HHAnalyzer::analyze(const edm::Event& event, const edm::EventSetup&, const
         if (allelectrons.p4[ielectron].Pt() > m_subleadingElectronPtCut
             && fabs(allelectrons.p4[ielectron].Eta()) < m_electronEtaCut) 
         {
+            // USEFUL?
```
good question...

swertz replied:
Not used at the moment...

OlivierBondu commented on interface/HHAnalyzer.h:
```diff
@@ -105,20 +109,21 @@ class HHAnalyzer: public Framework::Analyzer {
+        ONLY_NOMINAL_BRANCH(nMuonsM, unsigned int);
+        ONLY_NOMINAL_BRANCH(nElectrons, unsigned int);
+        //BRANCH(nElectronsL, unsigned int);
+        ONLY_NOMINAL_BRANCH(nElectronsT, unsigned int);
```
shouldn't it be the opposite, tight muons and medium electrons ?

blinkseb replied:
Duh, indeed...

OlivierBondu commented on plugins/HHAnalyzer.cc:
```diff
@@ -900,111 +880,66 @@ void HHAnalyzer::analyze(const edm::Event& event, const edm::EventSetup&, const
+            }
+
+            if (leptons[ilepton].isEl) {
                 nElectronsT++;
```
same here: tight muons and medium electrons for the counters :)

OlivierBondu commented on plugins/Tools.cc:
```diff
@@ -464,19 +480,8 @@ void HHAnalyzer::fillTriggerEfficiencies(const Lepton & lep1, const Lepton & lep
-    dilep.trigger_efficiency_downVariated_Arun = ((nominal - std::sqrt(error_squared_down_Arun)) < 0.)? 0. : nominal - std::sqrt(error_squared_down_Arun);
-
+    dilep.trigger_efficiency_upVariated_Arun = 0;
+    dilep.trigger_efficiency_downVariated_Arun = 0;
```
do we want to even keep these at all ? not harmful anyway...

blinkseb replied:
Let's remove :)

OlivierBondu reviewed (changes requested):
Looks good to me ! only a minor typo (tight muon and medium electron, not the other way around) :)

Commit:
Split JEC by sources and refactoring

blinkseb commented:
I updated my last commit

blinkseb commented:
And also fixed DZ filter efficiency application.

swertz commented on BuildFile.xml:
```diff
@@ -13,3 +13,4 @@
 <export>
     <lib name="1"/>
 </export>
+<flags CXXFLAGS="-g" />
```
I guess this was for you debugging and can be removed?

blinkseb replied:
It does not hurt the runtime, but does provide line number after a crash for I'd keep it :+1:

swertz commented on plugins/BuildFile.xml:
```diff
@@ -4,3 +4,4 @@
 <use name="cp3_llbb/Framework"/>
 <use name="cp3_llbb/TreeWrapper"/>
 <flags EDM_PLUGIN="1"/>
+<flags CXXFLAGS="-g" />
```
Same here?

swertz commented on plugins/HHAnalyzer.cc:
```diff
@@ -636,12 +611,14 @@ void HHAnalyzer::analyze(const edm::Event& event, const edm::EventSetup&, const
+
             myjet.CSV = alljets.getBTagDiscriminant(ijet, "pfCombinedInclusiveSecondaryVertexV2BJetTags");
             myjet.CMVAv2 = alljets.getBTagDiscriminant(ijet, "pfCombinedMVAV2BJetTags");
             float mybtag = alljets.getBTagDiscriminant(ijet, m_jet_bDiscrName);
```
`btag_L` and `btag_T` could be commented out, no?

swertz commented on plugins/HHAnalyzer.cc:
```diff
@@ -900,111 +876,67 @@ void HHAnalyzer::analyze(const edm::Event& event, const edm::EventSetup&, const
-        if (leptons[ilepton].id_L && leptons[ilepton].iso_L)
+    nJetsL = nJets;
+    if (! doingSystematics()) {
+        nBJetsL = 0;
```
Also here, for now we do not need L and T

swertz reviewed (commented):
Great thanks a lot! I do not see any glaring problems, but a few branches are now superfluous aren't they?

Commit:
Update HLT efficiencies for 2016 data

blinkseb commented:
One more commit with HLT efficiencies :+1: I also updated the refactoring to remove unneeded branches!

OlivierBondu commented:
good for me ! @swertz ?

swertz commented:
Still checking sth, sorry...

blinkseb commented:
L1 mitigation for muons is missing

blinkseb commented:
If you need to update some fixes, my branch is opened to every members of the cp3-llbb organization, so don't hesitate to push directly on it :)

Commit:
Efficiency for L1 EMTF muon bug

Commit:
Add HT

swertz commented:
I've added L1 bug mitigation, and also the computation of HT into the analyzer (easier than to recompute it in the plotter).

Commit:
Fix eta -> abs(eta)

OlivierBondu commented on plugins/Tools.cc:
```diff
@@ -364,6 +377,10 @@ void HHAnalyzer::fillTriggerEfficiencies(const Lepton & lep1, const Lepton & lep
         DZ_filter_eff = DZ_filter_eff_MuMu;
+        // L1 EMTF bug
+        if (getMuonsSector(lep1.p4) == getMuonsSector(lep2.p4) && sameEndcap(lep1.p4, lep2.p4)) {
+            DZ_filter_eff *= L1_EMTF_bug_eff_MuMu;
```
mmm, the L1_EMTF efficiency is averaged over the full run alreay, I guess we want to multiply it anyway, but reject altogether events affected by the bug ?

OlivierBondu replied:
my bad

swertz commented on test/HHConfiguration.py:
```diff
@@ -71,40 +71,34 @@
-
-framework.doSystematics(['jec', 'jer'])
+if not runOnData:
+    framework.smearJets(resolutionFile='cp3_llbb/Framework/data/Spring16_25nsV10_MC_PtResolution_AK4PFchs.txt', scaleFactorFile='cp3_llbb/Framework/data/Spring16_25nsV10_MC_SF_AK4PFchs.txt')
```
Sorry why do we keep `redoJEC()` and `smearJets()`? I thought the JEC had been updated in the GT?

blinkseb replied:
There's no smearing on miniaod so it has to be enabled for MC. For the JEC, I think there's been an update for MC, but I'm not sure. For the data it's not the last one in data (to be checked for the reminiaod).

swertz replied:
Ok thanks!

swertz replied:
It seems it was updated both in data and MC: https://hypernews.cern.ch/HyperNews/CMS/get/calibrations/2804.html

swertz commented:
So only two things to add to the PR: muon Rochester, EGM regression & scale... And then go!

OlivierBondu commented:
maybe we can merge this PR ? We're unlikely to forget these two items IMO...

swertz commented:
Ok, will do another one later then!
